My Site Dashboard: Extract NewBlogDetailHeaderView to SitePickerViewController

momo-ozawa opened the pull request:
Part of #17740

## Description
This PR hides the `NewBlogDetailHeaderView` in the table view belonging to `BlogDetailsViewController` and adds a `NewBlogDetailHeaderView` to`SitePickerViewController`.
- Embeds a `SitePickerViewController`, which contains a  `NewBlogDetailHeaderView` to the top of the stack view in `MySiteViewController`
- Hides the QuickActions row in the `NewBlogDetailHeaderView` displayed in `MySiteViewController`
- `SitePickerViewController` conforms to `BlogDetailHeaderViewDelegate`

Before | After
-- | --
<img src="https://user-images.githubusercontent.com/6711616/148992394-5aa9d42c-2e4d-4fdb-a14e-f0f51268e341.png" width=300> | <img src="https://user-images.githubusercontent.com/6711616/150147089-c59e9a87-9fbb-4ae6-8b29-e4150369b978.png" width=300>

## Notes
- These changes are behind a feature flag.
- There may be other logic that needs to be copied over from BlogDetailsVC in the future. For this PR, I've copied over only the logic needed to implement the `BlogDetailHeaderViewDelegate` to `SitePickerViewController`.

## How to review

To make sure there's no regressions in the `BlogDetailHeaderViewDelegate`, please review the implementation commits one by one by comparing it to the [implementation](https://github.com/wordpress-mobile/WordPress-iOS/blob/a08cc72e4bfb15d4fa88c06c549c16941fb2a64a/WordPress/Classes/ViewRelated/Blog/Blog%20Details/BlogDetailsViewController.m#L1097-L1165) in `BlogDetailsViewController`.

- ae1f72c68e3041aad0aebd72bd4a214c32db5a18 (siteSwitcherTapped)
- 6cc75d8254780f83a338025e62db0ea2f40a41af (siteTitleTapped)
- a29189b404258ba76e21f2d6176e1da798f6e934 (siteIconShouldAllowDroppedImages)
- cfec913f24269045c6f01d0860acdf874aa80984 (siteIconTapped)
- 10e0b1d49f65b3b357ac6a640b533cb1312d26ca (siteIconReceivedDroppedImage)
- ef5f233adeedbdf671a68a38bc859935231fcbf6 (visitSiteTapped)

## How to test

Make sure the My Site Dashboard feature flag is **enabled**.

**siteSwitcherTapped**
1. Go to My Site
2. Tap on the site switcher
3. Switch to a different site
4. ✅ Notice that the site switcher displays the selected site
5. ✅ Notice that the site menu (BlogDetailsVC) has updated to the selected blog

**siteTitleTapped**
1. Go to My Site
2. Tap on the site title
3. Change your site title
4. Tap save
5. ✅ Notice that your site title has been updated

**siteIconTapped**
1. Go to My Site
2. Tap on the site icon
3. Update your site icon
4. ✅ Notice that your site icon has been updated
5. Remove your site icon
6. ✅ Notice that your site icon has been removed

**visitSiteTapped**
1. Go to My Site
2. Tap on the site url in the header view
3. ✅ Notice that your site is displayed in a web vc

## Regression Notes
1. Potential unintended areas of impact
BlogDetailsHeaderView regression

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I tested the above flows manually

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: MySiteVC extension for header related logic

Commit:
Update: constrain stackview leading / trailing to superview

Previously the stackview was constrainted to the safe area.

Commit:
Add: set up the blog detail header view

Inserts a NewBlogDetailHeaderView as the very first element of the stack view, if the My Site Dashboard feature flag is turned on.

Commit:
Refactor: configure constraints for My Site Dashboard

The current NewBlogDetailsHeaderView displays Quick Actions as its subview.

In the My Site Dashboard project, Quick Actions will be displayed as a card view. This will be addressed in a future PR.

For now, this commit hides the Quick Actions row by:
- Not adding Quick Actions row as a subview / not activiating Quick Actions constraints  if the MSD feature flag is true
- Setting a bottom constraint for the title view if the MSD feature flag is true

Commit:
Add: make MySiteVC conform to  blog detail header view delegate

Commit:
Update: set blog for header view

Commit:
Refactor: inject blog service in constructor

Commit:
Refactor: implement site switcher tapped

This commit copies the sitSwitcherTapped() implementation from BlogDetailsViewController.

Commit:
Refactor: implement site title tapped

This commit copies the siteTitleTapped() implementation from BlogDetailsViewController.

Commit:
Refactor: implement site icon should allow dropped images

This commit copies the siteIconShouldAllowDroppedImages() implementation from BlogDetailsViewController.

Commit:
Refactor: implement site icon tapped

This commit copies the siteIconTapped() implementation from BlogDetailsViewController.

Commit:
Refactor: implement site icon received dropped image

This commit copies the siteIconReceivedDroppedImage() implementation from BlogDetailsViewController.

Commit:
Refactor: implement visitSiteTapped

This commit copies the visitSiteTapped() implementation from BlogDetailsViewController.

Commit:
Add: hide blog detail header on Blog Details VC

If the MySiteDashboard flag is true, the the header view in BlogDetailsViewController will not be shown or configured.

Only the header view in MySiteViewController will be visible.

Commit:
Add: SitePickerViewController

This vc displays the NewBlogDetailHeaderView, and is meant to be displayed as a child vc in MySiteViewController.

Commit:
Refactor: embed SitePickerViewController

Instead of directly adding a NewBlogDetailHeaderView to MySiteViewController, add SitePickerViewController (which contains a NewBlogDetailHeaderView) as a child view controller to MySiteViewController.

This allows most oftheheader view logic to be contained within the SitePickerViewController.

Commit:
Delete: MySiteViewController extensions

These extension files are now obsolete since the NewBlogDetailHeaderView was extracted to SitePickerViewController.

momo-ozawa commented:
@leandroalonso Ready for review -- extracted the `NewBlogDetailHeaderView` to `SitePickerViewController`, which is then embedded as a child vc in `MySiteViewController`

pachlava commented:
Hey @momo-ozawa, thanks for providing the detailed tests with the PR! I did some testing on my side, and so far I have only one really minor thing. Maybe it's planned for rework anyway. There's an empty space at the top of the Site Menu, which looked OK with the old layout, but after the rearrangement it becomes more obvious, especially under the dark theme:

![photo_2022-01-20 12 14 30](https://user-images.githubusercontent.com/73365754/150319462-4d70e686-e1ef-4961-866e-a1fa6036522f.jpeg)

Here's with the light theme:

![photo_2022-01-20 12 08 27](https://user-images.githubusercontent.com/73365754/150319690-1c8be32b-b8ab-4711-b823-0e313a43c459.jpeg)

It looks like according to designs it should not be there. However, when removed, the whole screen will look a bit off under light theme, as long as the background for the segmented control remains white (it looks like it should be grey, from designs, as well as the background for the site title).

momo-ozawa commented:
@pachlava 👋 

Thanks for your feedback! This PR is more of an architectural PR, and design refinements will be addressed later once the foundational PR's have landed 🙏

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:
```diff
@@ -105,6 +107,24 @@ class MySiteViewController: UIViewController, NoResultsViewHost {
         NotificationCenter.default.addObserver(self, selector: #selector(showAddSelfHostedSite), name: .addSelfHosted, object: nil)
     }
 
+    private func showSitePicker(for blog: Blog) {
```
There's an issue with displaying the site picker for the following flow:

1. Turn the My Site Dashboard flag to on
2. Log out
3. Login again and select a site
4. Tap "Show me around"
5. No site picker is shown

momo-ozawa replied:
Fixed! 7884be2

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Site Picker/SitePickerViewController+QuickStart.swift:
```diff
@@ -0,0 +1,34 @@
+private var alertWorkItem: DispatchWorkItem?
+private var observer: NSObjectProtocol?
+
+extension SitePickerViewController {
```
I tested QuickStart with this new `SitePickerViewController` and the glowing signal is not appearing at all. This is how it is with the flag disabled:

<img width="384" alt="Screen Shot 2022-01-20 at 12 35 03" src="https://user-images.githubusercontent.com/7040243/150370645-650dd91b-0407-46a5-8bb1-e3cecab1280b.png">

I'm still able to complete the quick start steps, though.

momo-ozawa replied:
Fixed! 72d2296

leandroalonso reviewed (commented):
@momo-ozawa I tested different flows and compared the current behavior with the current `trunk`.

There were a few issues that I found that are shared in this review. I also double-checked the events to make sure we're not removing any of them by mistake.

leandroalonso commented:
@momo-ozawa another thing that I noticed is that pull to refresh doesn't update site details anymore (made on another device).

Commit:
Fix: configure site picker when blog is set

Commit:
Fix: start observing quick start

This fixes an issue where the spotlight wasn't being toggled for the site title and the site icon tours.

Commit:
Fix: whitespace

momo-ozawa commented:
@leandroalonso  👋 

> @momo-ozawa another thing that I noticed is that pull to refresh doesn't update site details anymore (made on another device).

I'll address this in a different PR, when I work on [this issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/17784) 

I've addressed the other issues you found -- ready for a re-review! 🙇‍♀️ 

https://user-images.githubusercontent.com/6711616/150538469-7921ff0e-5b5d-425a-af7a-75be3ee86566.mp4

leandroalonso reviewed (approved):
@momo-ozawa tested all the flows, all good! :shipit:
